includes/amp-helper-functions.php

```diff
@@ -11,12 +11,28 @@ function amp_get_permalink( $post_id ) {
 	if ( empty( $structure ) ) {
 		$amp_url = add_query_arg( AMP_QUERY_VAR, 1, get_permalink( $post_id ) );
 	} else {
-		$amp_url = trailingslashit( get_permalink( $post_id ) ) . user_trailingslashit( AMP_QUERY_VAR, 'single_amp' );
+		$amp_url = amp_normalize_url_param( get_permalink( $post_id ) );
 	}
 
 	return apply_filters( 'amp_get_permalink', $amp_url, $post_id );
 }
 
+/**
+ * For normalizing urls with parameters
+ */
+function amp_normalize_url_param( $url ) {
+	if ( false === strpos( $url, '?' ) ) {
+		return trailingslashit( $url ) . user_trailingslashit( AMP_QUERY_VAR, 'single_amp' );
+	}
+
+	$parts   = explode( '?', $url, 2 );
```

Comment:
Instead of exploding on `?`, this should use `wp_parse_url()` to parse out the URL parts and then either amend the query string with the `AMP_QUERY_VAR` param or amend the `AMP_QUERY_VAR` slug to the path.

Reply:
For example, it is possible that `?` appears in a fragment identifier: `http://example.com/#/?foo/bar`